File: 0017_letter_combinations_of_a_phone_number.py
# https://leetcode-cn.com/problems/letter-combinations-of-a-phone-number/

class Solution:
    def letterCombinations(self, digits):
        """
        Inputs:
            digits: str
        Outputs:
            res: list[str]
        """
        
        '''
        if not digits:
            return []
        
        # Digit-string map.
        m = ['', '*', 'abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz']
        res = []
        
        def backtracking(tmp, idx):
            if idx == len(digits):
                res.append(tmp)
                return
            
            digit = digits[idx] # current digit
            # Letters corresponding to the current digit.
            letters = m[ord(digit) - ord('0')]
            
            for i in letters:
                backtracking(tmp+i, idx+1)
        
        backtracking('', 0)
        return res
        '''
        
        if not digits:
            return []
        
        # Digit-string map.
        m = ['', '*', 'abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz']
        res = ['']
        for i in digits:
            # Letters corresponding to the current digit.
            letters = m[ord(i) - ord('0')]
            for _ in range(len(res)):
                tmp = res.pop(0)
                for j in letters:
                    res.append(tmp+j)
        return res

File: Cpp/0017_letter_combinations_of_a_phone_number.cpp
// https://leetcode-cn.com/problems/letter-combinations-of-a-phone-number/

class Solution {
public:
    vector<string> letterCombinations(string digits) {
        unordered_map<char, string> table{
            {'2', "abc"}, {'3', "def"},
            {'4', "ghi"}, {'5', "jkl"}, {'6', "mno"},
            {'7', "pqrs"}, {'8', "tuv"}, {'9', "wxyz"}
        };
        vector<string> res;
        if (digits.size() == 0)
            return res;

        res.push_back("");
        for (char num: digits) {
            int res_len = res.size();
            for (int i = 0; i < res_len; ++i) {
                string tmp = res.back();
                res.pop_back();

                for (char c: table[num]) {
                    string tmp_new = tmp + c;
                    res.insert(res.begin(), tmp_new);
                }
            }
        }

        return res;
    }
};